src/server/server_family.cc

```diff
@@ -1763,18 +1763,39 @@ std::string ServerFamily::BuildClusterNodeReply(ConnectionContext* cntx) const {
   }
 }
 
+template <class Lock> bool try_lock_for(Lock& lock, absl::Duration duration) {
+  absl::Time start = absl::Now();
+  while (true) {
+    VLOG(1) << "Try acquire lock";
+    if (lock.try_lock()) {
+      VLOG(1) << "Acquired lock";
+      return true;
+    }
+    if (absl::Now() - start > duration) {
+      return false;
+    }
+    ThisFiber::SleepFor(200ms);
+  }
+}
+
 void ServerFamily::ReplicaOf(CmdArgList args, ConnectionContext* cntx) {
   std::string_view host = ArgS(args, 1);
   std::string_view port_s = ArgS(args, 2);
   auto& pool = service_.proactor_pool();
 
   LOG(INFO) << "Replicating " << host << ":" << port_s;
 
-  if (absl::EqualsIgnoreCase(host, "no") && absl::EqualsIgnoreCase(port_s, "one")) {
-    // use this lock as critical section to prevent concurrent replicaof commands running.
-    VLOG(1) << "Acquire replica lock";
-    unique_lock lk(replicaof_mu_);
+  // From here until the end of Replica::Start() it's a critical section. We don't want to wait
+  // for the lock too long because that hangs the client, so we try to acquire it for ~2 seconds and
+  // return an error if we can't. We should eventually break up the critical section so we don't
+  // have to hold the lock during the networking parts.
+  unique_lock lk(replicaof_mu_, std::defer_lock);
+  if (!try_lock_for(lk, absl::Seconds(2))) {
+    (*cntx)->SendError("Another replication command in progress");
+    return;
+  }
 
```

Comment:
@adiholden 

I have another suggestion. If the mutex is locked, we force-cancel the replication and wait until we can lock again. And then proceed with the new REPLICAOF command. I guess this is what is expected by users?

For the replica, cancellation should be reliable - stop without any delays and deadlocks. So we don't even need the timed wait 🤔

Reply:
Is there a way to cancel a `Connect()` call? If not then this flow can still block for an annoying amount of time.

Reply:
Yes, the replica has built-in cancellation, that closes sockets. I'm not sure it works right away for connect, but it does for later stages.

You can look at replica->Stop and all the stuff with the context. Its a little difficult unfortunately. There should also be some comments around it and the Context in common.h

We can call replica->Stop without any mutex, so it can guard only the pointer

Reply:
I agree we should cancel the the replication process and wait for the mutex to until we can lock it for the new replicaof command.

Reply:
I changed the locking/cancelling flow. please take another look :)